src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -215,24 +218,45 @@ private Optional<ThingAssociations> getThingsAssociatedWithCoreDevice() {
         return Optional.empty();
     }
 
-    private void refresh(ThingAssociations associations) {
-        associations.getThings().forEach(this::refreshCertificateThingAttachment);
-        associations.getStaleThings().forEach(thingRegistry::deleteThing);
-
-        certificateRegistry.getAllCertificates().forEach(certificate -> {
-            if (associations.isAttachedToThing(certificate)) {
-                refreshCertificateValidity(certificate);
-            } else {
-                certificateRegistry.deleteCertificate(certificate);
-            }
-        });
+    private void refresh(Stream<Thing> thingsAttachedToCore) {
+        Pair<Set<String>, Consumer<Thing>> refreshCertificates = refreshCertificateValidity();
+
+        Set<String> attachedThingNames = thingsAttachedToCore
+                // Refresh the thing certificate attachments
+                .peek(this::refreshCertificateAttachments)
```

Comment:
in general we should avoid using peek to do real work, as it [exists mainly to support debugging](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-), and has some [pitfalls](https://rules.sonarsource.com/java/RSPEC-3864).